### ISSUE TITLE
Design: flag to delimit `onyo get` pathnames with null (`\0`)

### ISSUE THREAD

**Issue by aqw** [psyinfra/onyo#686]
It is considered best-unix-practice to delimit output with `\0`, especially when the contents could contain problematic characters (spaces, backslashes, tabs, etc). 

It is especially useful with `xargs` (from the manpage):

> `-0`, `--null`
> Input  items  are terminated by a null character instead of by whitespace, and the quotes and backslash are not special (every character is taken literally).  Disables the end-of-file string, which is treated like any other argument.  Useful when input items might contain white space, quote marks, or backslashes.  The GNU find `-print0` option produces input suitable for this mode.

and

> [...] The `-d` option disables any end-of-file string, which is treated like any other argument.  You can use this option when the input consists of simply newline-separated items, although it is almost always better to design your program to use `--null` where this is possible.

### Naming

There is... little agreement on the name of such a flag:

- `find`: `-print0`
- `git grep`: `-z, --null`
- `grep`: `-Z, --null`
- `xargs`: `-0, --null`

My preference is `-z, --null` (though I am also ok with `-0, --null`).

### Delimiting

This is where things get interesting.

The primary use-case is:
```
onyo get --keys path --null
```

`find -print0` outputs paths and delimits them with `\0`. We should match exactly that.

However, `onyo get` can output more than just the `path` key.  `grep` (and `git grep`) mix paths and output, so they've already explored this ground. From `git grep`'s manpage:

> Use `\0` as the delimiter for pathnames in the output [...]

And indeed, they both do exactly that:

```
pathname\0matching-text\n
```

So the question is, what should be the output of `onyo get --keys path make model.name --null`:
1) `path\0make\tmodel.name\n` (aka: delimit the `path`)
2) `path\tmake\model.name\0` (aka: delimit the line)

The advantage of Option 1 is that it matches `grep`'s behavior, and one can easily determine the `path` in the output (even if `path` is in a different column). If one puts `path` in the first column, then there is no ambiguity in the case that a pathname contains `\t`.

Option 2 has the advantage that there are no special cases. All columns are treated the same no matter what. `\n` is never used. If one only queries for `path`, there is no ambiguity in the case that a pathname contains `\t`. Plus, it also makes the feature useful for `path.directory`, `path.file`, etc (see #678).

I favor Option 2. Users only have primitive control over `grep`'s output, but with `onyo get` they create the query. That, and the other keys that can be queried, I think Option 2 is more powerful and a better match for how people may use/abuse `onyo get`.

### Characters

`git grep` does something interesting, and takes `--null` a step further; it changes how paths are output in `git grep`.

> `-z, --null`
> Use `\0` as the delimiter for pathnames in the output, and print them verbatim. Without this option, pathnames with "unusual" characters are quoted as explained for the configuration variable `core.quotePath` (see git-config(1)).

The relevant part of the `git-config` manpage:

> `core.quotePath`
> Commands that output paths (e.g.  `ls-files`, `diff`), will quote "unusual" characters in the pathname by enclosing the pathname in double-quotes and escaping those characters with backslashes in the same way C escapes control characters (e.g.  `\t` for TAB, `\n` for LF, `\\` for backslash) or bytes with values larger than 0x80 (e.g. octal \302\265 for "micro" in UTF-8). If this variable is set to `false`, bytes higher than 0x80 are not considered "unusual" any more. Double-quotes, backslash and control characters are always escaped regardless of the setting of this variable. A simple space character is not considered "unusual". Many commands can output pathnames completely verbatim using the `-z` option. The default value is `true`.

Onyo already prints the character for values larger than 0x80, and I see no value in changing that behavior. However, control characters are an open question:

I tested with a tab character in the `type` field:
```
❱ onyo get --path shelf -H           
desk	top	dell	D11s	abc123	shelf/desk	top_dell_D11s.abc123
...
❱ git ls-files shelf 
...
"shelf/desk\ttop_dell_D11s.abc123"
...
❱ git ls-files -z shelf
[...]shelf/desk	top_dell_D11s.abc123[...]
```

So the question is, should `onyo`:

1) disallow control characters in keys/values
2) match `git grep`'s behavior (quote and escape normally, literal when in `--null`)
3) current behavior (don't quote, don't escape ever)

I must admit that my initial impulse is option 1, but that eliminates long text as a value (which we likely will never do, but is a valid use-case).

I am deeply ambivalent (or rather, amtrivalent ;-).

The outcome of this should become its own issue.

### Machine Readable

I suggest that `--null` also set `--machine-readable` automatically. This mode is only useful for parsing, and I find it difficult to even imagine how `--null` would behave with the decoration, much less a use-case.

**Comment by bpoldrack** [comment]
> I favor Option 2. Users only have primitive control over grep's output, but with onyo get they create the query.

I agree.

> I suggest that --null also set --machine-readable automatically.

Agree on this as well.

Re characters:

> So the question is, should onyo:
>
>  1. disallow control characters in keys/values
>  2. match git grep's behavior (quote and escape normally, literal when in --null)
>  3. current behavior (don't quote, don't escape ever)

I don't like 1. "long" text is absolutely valid and we should support it.
I'd lean towards 2. with the clarification that this is relevant for `--machine-readable`. When we (rich-)print stuff, we don't escape, I think. If someone sees how I am wrong in that, point it out, please.

**Comment by aqw** [comment]
> So the question is, should onyo:
>
>  1. disallow control characters in keys/values
>  2. match git grep's behavior (quote and escape normally, literal when in --null)
>  3. current behavior (don't quote, don't escape ever)

I agree that 1 is a valid use-case. So we need to decide "something" about whether we escape, or not, or allow different delimiters.

If we allow control chars (specifically tab) in values, then we have an issue with `onyo get`, because we hardcode `\t` as our delimiter.

If we escape the values, then it's not in a format that's ready-to-use for scripting. If we don't escape it, then scripts can't iterate over multiple fields.

I don't think onyo can make a one-size-fits-all right decision here. The behaviors I see are:
1) `--delimiter-assets` (default '\n'; `-z, --null` would be a shorthand for `--delimiter-assets='\0' --machine-readable`)
2) `--delimiter-values` (default `\t'; users could override to something they feel pretty sure is not in their data)
3) `--escape-values` (default `False`; users can use this flag to escape the output)

I don't like that this turns into a feature flag orgy.

IMO, the least useful is `--delimiter-values`. If someone has tabs in their data, then I struggle to imagine what character they won't have. Newlines? Commas? Pipes? I think at that point, they have no choice but to operate on escaped data, and process it.

Thoughts?

**Comment by aqw** [comment]
So, this is basically all solved, except for escaping [control characters](https://en.wikipedia.org/wiki/Control_character). I propose the following:

1) disallow any control characters in key names (add fsck check, and preferably schema)
2) disallow any control characters in the values of keys that are used in the asset name (i.e. no control characters in asset names)
    - this protects/solves the 99% use-case
3) control characters in any other value are allowed, but will be escaped when printed

This does mean we will disallow `BEL`, which is a loss, but one I'll live with. ;-)

Open Question:
- do we follow `git`'s lead and not escape control characters in `--null` mode? I don't see the advantage right now, but am very interested in use-cases if someone has one.

**Comment by bpoldrack** [comment]
Agreed. 

> do we follow git's lead and not escape control characters in --null mode? I don't see the advantage right now, but am very interested in use-cases if someone has one.

I think the point is, that it's likely not intended for actual printing but piped into something that parses the literal binary.
One doesn't need `--null` when printing to a terminal.
So, I lean towards following git.

Other than that: I think it would be nice to allow choosing a different value delimiter. If we want to avoid option spam, we could make it a config. Keeping in mind that eventually we'd want to allow env vars to override configs (which then gives a way to set this per-call w/o adding yet another option).

**Comment by aqw** [comment]
> I think the point is, that it's likely not intended for actual printing but piped into something that parses the literal binary.
> One doesn't need --null when printing to a terminal.

Yeah, the consequence of this is that if I query multiple fields with `onyo get`, and one of those fields contains unescaped control characters (lets say "tab"), then my field splitting is broken for whatever I pipe into. Escaping protects that.

> Other than that: I think it would be nice to allow choosing a different value delimiter. If we want to avoid option spam, we could make it a config. Keeping in mind that eventually we'd want to allow env vars to override configs (which then gives a way to set this per-call w/o adding yet another option).

 Big +1 on allowing this to be a configuration option and env var (we should open an issue for that). I do want to be more careful about adding option flags. That said, this one might be important enough to warrant being a flag.

**Comment by aqw** [comment]
I decided to take a look at how other utilities solve this. `find` has a section in the manpage called [Unusual Filenames](https://manpages.debian.org/bookworm/findutils/find.1.en.html#UNUSUAL_FILENAMES).

At least on an initial read, I find their solution to be unhelpful.

The one thing I take away from this is that they seem to be more concerned with the terminal interpreting the control characters, rather than word splitting.
